sos/plugins/juju.py

```diff
@@ -37,85 +19,59 @@ class Juju(Plugin, UbuntuPlugin):
-        ('export-mongodb',
-         'Export mongodb collections as json files', '', False),
-        ('generate-bundle',
-         """Generate a YAML bundle of the current environment
-         (requires juju-deployerizer)""", '', False),
-    ]
-
-    def get_deployed_services(self):
-        cmd = "juju status --format json"
-        status_json = self.call_ext_prog(cmd)['output']
-        self.add_string_as_file(status_json, "juju_status_json")
-        # if status_json isn't in JSON format (i.e. 'juju' command not found),
-        # or if it does not contain 'services' key, return empty list
-        try:
-            return json_loads(status_json)['services'].keys()
-        except ValueError:
-            return []
-
-    @ensure_service_is_running("juju-db")
-    def export_mongodb(self):
-        collections = (
-            "relations",
-            "environments",
-            "linkednetworks",
-            "system",
-            "settings",
-        )
-
-        for collection in collections:
-            self.add_cmd_output(
-                "/usr/lib/juju/bin/mongoexport --ssl \
-                --dbpath=/var/lib/juju/db --db juju --collection {0} \
-                --jsonArray".format(collection),
-                suggest_filename="{}.json".format(collection))
+    # Using files instead of packages here because there is no identifying
+    # package on a juju machine.
+    files = ('/var/log/juju')
 
     def setup(self):
-        self.add_copy_spec("/var/log/upstart/juju-db.log")
-        self.add_copy_spec("/var/log/upstart/juju-db.log.1")
-        if not self.get_option("all_logs"):
-            # We need this because we want to collect to the limit of all
-            # *.logs in the directory.
-            if(os.path.isdir("/var/log/juju/")):
-                for filename in os.listdir("/var/log/juju/"):
-                    if filename.endswith(".log"):
-                        fullname = os.path.join("/var/log/juju/" + filename)
-                        self.add_copy_spec(fullname)
-            self.add_cmd_output('ls -alRh /var/log/juju*')
-            self.add_cmd_output('ls -alRh /var/lib/juju/*')
+        # Get juju services, they are uniquely named based on the deployment
+        # so we need to parse what their names are.
+        cmd = "systemctl --no-pager --all --no-legend list-units juju\\*"
+        juju_systemctl_output = self.exec_cmd(cmd)["output"]
+        for line in iter(juju_systemctl_output.splitlines()):
+            match = re.match(r"^(juju\S+)\.", line)
+            if match:
+                self.add_journal(units=str(match.group(1)))
+                self.add_service_status(str(match.group(1)))
+
+        # Get agent configs for each agent.
+        agent_configs = []
+        for dirpath, dirnames, filenames in os.walk("/var/lib/juju/agents"):
+            for dirname in dirnames:
+                agent_configs.append(
+                    "/var/lib/juju/agents/" + dirname + "/agent.conf")
```

Comment:
This can all be reduced to:
`self.add_copy_spec("/var/lib/juju/agents/*/agent.conf"`)

Reply:
Fixed on https://github.com/sosreport/sos/pull/1935/commits/45b69234af2f42292794acf36f5f264c4b5f5105